### ISSUE TITLE
Failure to execute code copied from Multi-line hints using the Copy to clipboard feature.

### ISSUE THREAD

**Issue by cswclui** [rstudio/learnr#691]
Failure to execute code copied from Multi-line hints using the `Copy to clipboard` feature.

Consider the following sample learnr tutorial running in Ubuntu 20.04.3.

<pre>
---
title: "Test"
output: 
  learnr::tutorial:
    progressive: true
runtime: shiny_prerendered
---

```{r setup, include=FALSE,message=FALSE}
library("learnr")

new_recorder <- function(tutorial_id, tutorial_version, user_id, event, data) {
    cat("-------event received----\n")
    print(data$code)
    cat("-------end event ---\n\n-")
}
options(tutorial.event_recorder = new_recorder)
```

# Test

```{r demo, exercise = TRUE}
3
```

```{r demo-hint, eval=FALSE}
1
2
```

</pre>

Step to reproduce:
1. Click the `hint` button
2. Click `Copy from clipboard`
3. Paste code into the exercise box.

Actual output:
![image](https://user-images.githubusercontent.com/218137/167232149-54bb0115-f48d-4dc8-861e-815e543516f8.png)

It seems the problem is caused by the extra "\r" from multi-line hint when using the `Copy to clipboard` function. From the event object,  data$code has the value "1\r\n2". When typing the code manually in the exercise box (which can be executed), data$code has the value "1\n2\n".

**Comment by cswclui** [comment]
Your script is working on my computer, but it's not when I tried to add another section (Extra test) where I copy the hints from the second section. 

![image](https://user-images.githubusercontent.com/218137/167455028-c19d84b6-8eac-4f64-9013-c6bb53cf1bc9.png)

I'm wonder if my Rstudio (running in Linux) is adding an extra '\r' character. 
![image](https://user-images.githubusercontent.com/218137/167455637-c0b1ad3b-c66d-4c96-b800-a25a3f1de7f6.png)

Please refer to the following Rmd and html document.
[learnr_sample.zip](https://github.com/rstudio/learnr/files/8653575/learnr_sample.zip)

**Comment by gadenbuie** [comment]
That might be it! RStudio has a line ending conversion setting in Global Options > Code > Saving. Can you check what setting you have there?

![image](https://user-images.githubusercontent.com/5420529/167472172-235559d0-7419-46b1-a066-6c3d80fb5d51.png)

**Comment by cswclui** [comment]
> That might be it! RStudio has a line ending conversion setting in Global Options > Code > Saving. Can you check what setting you have there?
> 
> ![image](https://user-images.githubusercontent.com/5420529/167472172-235559d0-7419-46b1-a066-6c3d80fb5d51.png)

The default setting is "Native". I tried to change to other options but the same problem occurs.

**Comment by gadenbuie** [comment]
And setting it to POSIX doesn't change the behavior? That's very odd.  Can you please provide the output from `devtools::session_info(info = "all")`. That will help me consider next steps.

**Comment by cswclui** [comment]
This is the output of 
```devtools::session_info(info = "all")```

```
─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.1.2 (2021-11-01)
 os       Ubuntu 20.04.3 LTS
 system   x86_64, linux-gnu
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Etc/UTC
 date     2022-05-10
 pandoc   2.14.0.3 @ /usr/lib/rstudio-server/bin/pandoc/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package     * version     date (UTC) lib source
 backports     1.4.1       2021-12-13 [2] RSPM (R 4.1.0)
 brio          1.1.3       2021-11-30 [2] RSPM (R 4.1.0)
 bslib         0.3.1       2021-10-06 [2] RSPM (R 4.1.0)
 cachem        1.0.6       2021-08-19 [2] RSPM (R 4.1.0)
 callr         3.7.0       2021-04-20 [2] RSPM (R 4.1.0)
 checkmate     2.1.0       2022-04-21 [2] CRAN (R 4.1.2)
 cli           3.3.0       2022-04-25 [2] RSPM (R 4.1.0)
 crayon        1.5.1       2022-03-26 [2] RSPM (R 4.1.0)
 curl          4.3.2       2021-06-23 [2] RSPM (R 4.1.0)
 desc          1.4.1       2022-03-06 [2] RSPM (R 4.1.0)
 devtools      2.4.3       2021-11-30 [2] RSPM (R 4.1.0)
 digest        0.6.29      2021-12-01 [2] RSPM (R 4.1.0)
 ellipsis      0.3.2       2021-04-29 [2] RSPM (R 4.1.0)
 evaluate      0.15        2022-02-18 [2] RSPM (R 4.1.0)
 fastmap       1.1.0       2021-01-25 [2] RSPM (R 4.1.0)
 fs            1.5.2       2021-12-08 [2] RSPM (R 4.1.0)
 glue          1.6.2       2022-02-24 [2] RSPM (R 4.1.0)
 htmltools     0.5.2       2021-08-25 [2] RSPM (R 4.1.0)
 htmlwidgets   1.5.4       2021-09-08 [2] RSPM (R 4.1.0)
 httpuv        1.6.5       2022-01-05 [2] RSPM (R 4.1.0)
 jquerylib     0.1.4       2021-04-26 [2] RSPM (R 4.1.0)
 jsonlite      1.8.0       2022-02-22 [2] RSPM (R 4.1.0)
 knitr         1.39        2022-04-26 [2] RSPM (R 4.1.0)
 later         1.3.0       2021-08-18 [2] RSPM (R 4.1.0)
 learnr        0.10.1.9022 2022-05-07 [1] Github (rstudio/learnr@802e4ec)
 lifecycle     1.0.1       2021-09-24 [2] RSPM (R 4.1.0)
 magrittr      2.0.3       2022-03-30 [2] RSPM (R 4.1.0)
 markdown      1.1         2019-08-07 [2] CRAN (R 4.1.2)
 memoise       2.0.1       2021-11-26 [2] RSPM (R 4.1.0)
 mime          0.12        2021-09-28 [2] RSPM (R 4.1.0)
 pkgbuild      1.3.1       2021-12-20 [2] RSPM (R 4.1.0)
 pkgload       1.2.4       2021-11-30 [2] RSPM (R 4.1.0)
 prettyunits   1.1.1       2020-01-24 [2] RSPM (R 4.1.0)
 processx      3.5.3       2022-03-25 [2] RSPM (R 4.1.0)
 promises      1.2.0.1     2021-02-11 [2] RSPM (R 4.1.0)
 ps            1.7.0       2022-04-23 [2] RSPM (R 4.1.0)
 purrr         0.3.4       2020-04-17 [2] RSPM (R 4.1.0)
 R6            2.5.1       2021-08-19 [2] RSPM (R 4.1.0)
 Rcpp          1.0.8.3     2022-03-17 [2] RSPM (R 4.1.0)
 remotes       2.4.2       2021-11-30 [2] CRAN (R 4.1.2)
 rlang         1.0.2       2022-03-04 [2] RSPM (R 4.1.0)
 rmarkdown     2.14        2022-04-25 [2] RSPM (R 4.1.0)
 rprojroot     2.0.3       2022-04-02 [2] RSPM (R 4.1.0)
 rstudioapi    0.13        2020-11-12 [2] RSPM (R 4.1.0)
 sass          0.4.1       2022-03-23 [2] RSPM (R 4.1.0)
 sessioninfo   1.2.2       2021-12-06 [2] RSPM (R 4.1.0)
 shiny       * 1.7.1       2021-10-02 [2] RSPM (R 4.1.0)
 stringi       1.7.6       2021-11-29 [2] RSPM (R 4.1.0)
 stringr       1.4.0       2019-02-10 [2] RSPM (R 4.1.0)
 testthat      3.1.4       2022-04-26 [2] RSPM (R 4.1.0)
 usethis       2.1.5       2021-12-09 [2] RSPM (R 4.1.0)
 withr         2.5.0       2022-03-03 [2] RSPM (R 4.1.0)
 xfun          0.30        2022-03-02 [2] RSPM (R 4.1.0)
 xtable        1.8-4       2019-04-21 [2] RSPM (R 4.1.0)
 yaml          2.3.5       2022-02-21 [2] RSPM (R 4.1.0)

 [1] /home/rstudio/.R
 [2] /usr/local/lib/R/site-library
 [3] /usr/local/lib/R/library

─ External software ──────────────────────────────────────────────────────────
 setting        value
 cairo          1.16.0
 cairoFT
 pango          1.44.7
 png
 jpeg
 tiff
 tcl
 curl           7.68.0
 zlib           1.2.11
 bzlib          1.0.8, 13-Jul-2019
 xz             5.2.4
 PCRE           10.34 2019-11-21
 ICU            66.1
 TRE            TRE 0.8.0 R_fixes (BSD)
 iconv          glibc 2.31
 readline       8.0
 BLAS           /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
 lapack         /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
 lapack_version 3.9.0

─ Python configuration ───────────────────────────────────────────────────────
 Python is not available

──────────────────────────────────────────────────────────────────────────────
```

**Comment by cswclui** [comment]
I have built a docker image to demonstrate the issue. 

In **Windows OS**, 
1. Execute `docker run -p 3838:3838 vlabpolyu/learnrissue` in command prompt
2. In chrome browser, visit http://localhost:3838/sample.Rmd
3. Copy hint from the `Solution` section to the exercise block in the `Test` section

Sample output:
![image](https://user-images.githubusercontent.com/218137/167525048-7bcf333d-c557-46c2-b6d5-187a07f64e76.png)

**Comment by gadenbuie** [comment]
ohhh, okay, this is starting to make some sense now. I believe the issue originates in Windows, where `\r\n` is the standard line-ending indicator. In Posix (unix and Mac) the line-ending is just `\n`. My guess is that when you copy the text in Windows, it includes the `\r\n` line endings, which are not expected by the Unix system running in Docker. I'm a little surprised that the RStudio IDE doesn't clean that up when pasting, but maybe that's because the input comes from a browser (in Windows)?

I need to do a little investigation to see if it would be acceptable to always convert from CRLF line endings `\r\n` to the Posix `\n` ending. Alternatively, learnr might need to be more careful about character encoding, or we might need to resort to a more complicated process.

**Comment by cswclui** [comment]
I suppose it's related to clipboard copying button. The bug is triggered when I use the `copy to clipboard` button in Windows (it will pate the \r\n). There is no problem when I am selecting the text manually and use Ctrl+C to copy and paste it.

> ohhh, okay, this is starting to make some sense now. I believe the issue originates in Windows, where `\r\n` is the standard line-ending indicator. In Posix (unix and Mac) the line-ending is just `\n`. My guess is that when you copy the text in Windows, it includes the `\r\n` line endings, which are not expected by the Unix system running in Docker. I'm a little surprised that the RStudio IDE doesn't clean that up when pasting, but maybe that's because the input comes from a browser (in Windows)?
> 
> I need to do a little investigation to see if it would be acceptable to always convert from CRLF line endings `\r\n` to the Posix `\n` ending. Alternatively, learnr might need to be more careful about character encoding, or we might need to resort to a more complicated process.

**Comment by gadenbuie** [comment]
Ah, okay, I see the problem now. When you click the Copy to Clipboard button, we request the text in the editor from Ace, the JavaScript library that powers the editor. Because it detects that you're on Windows, it uses the CRLF line endings. But what really matters is the server OS – and even then it's possible that R's parser will handle POSIX line endings in R sessions on Windows. Anyway, I have a good understanding of the issue now; I'll try to put together a fix tomorrow.

**Comment by gadenbuie** [comment]
Hi @cswclui, would you mind trying out the fix in #692 to see if that resolves your issue?

```r
devtools::install_github("rstudio/learnr#692")
```

**Comment by cswclui** [comment]
> Hi @cswclui, would you mind trying out the fix in #692 to see if that resolves your issue?
> 
> ```r
> devtools::install_github("rstudio/learnr#692")
> ```

The problem yes fixed on my side.
Thank you very much.

**Comment by gadenbuie** [comment]
Thanks for reporting and helping test the fix, @cswclui!
